optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/list/ListSwapMove.java

```diff
@@ -0,0 +1,216 @@
+    private final Object leftEntity;
+    private final int leftIndex;
+    private final Object rightEntity;
+    private final int rightIndex;
+
+    /**
+     * Create a move that swaps a list variable element at {@code leftEntity.listVariable[leftIndex]} with
+     * {@code rightEntity.listVariable[rightIndex]}.
+     *
+     * <h4>ListSwapMove anatomy</h4>
+     *
+     * <pre>
+     * {@code
+     *                                 ┌ rightEntity
+     *                right element ┐  │   ┌ rightIndex
+     *                              ↓  ↓   ↓
+     *               A {Ann[0]} <-> Y {Bob[1]}
+     *               ↑  ↑   ↑
+     *  left element ┘  │   └ leftIndex
+     *                  └ leftEntity
+     * }
+     * </pre>
+     *
+     * <h4>Example</h4>
+     *
+     * <pre>
+     * {@code
+     * GIVEN
+     * Ann.tasks = [A, B, C]
+     * Bob.tasks = [X, Y]
+     *
+     * WHEN
+     * ListSwapMove: A {Ann[0]} <-> Y {Bob[1]}
+     *
+     * THEN
+     * Ann.tasks = [Y, B, C]
+     * Bob.tasks = [X, A]
+     * }
+     * </pre>
+     *
+     * @param variableDescriptor descriptor of a list variable, for example {@code Employee.taskList}
+     * @param leftEntity together with {@code leftIndex} identifies the left element to be moved
+     * @param leftIndex together with {@code leftEntity} identifies the left element to be moved
+     * @param rightEntity together with {@code rightIndex} identifies the right element to be moved
+     * @param rightIndex together with {@code rightEntity} identifies the right element to be moved
+     */
+    public ListSwapMove(
+            ListVariableDescriptor<Solution_> variableDescriptor,
+            Object leftEntity, int leftIndex,
+            Object rightEntity, int rightIndex) {
+        this.variableDescriptor = variableDescriptor;
+        this.leftEntity = leftEntity;
+        this.leftIndex = leftIndex;
+        this.rightEntity = rightEntity;
+        this.rightIndex = rightIndex;
+    }
+
+    public Object getLeftEntity() {
+        return leftEntity;
+    }
+
+    public int getLeftIndex() {
+        return leftIndex;
+    }
+
+    public Object getRightEntity() {
+        return rightEntity;
+    }
+
+    public int getRightIndex() {
+        return rightIndex;
+    }
+
+    public Object getLeftValue() {
+        return variableDescriptor.getElement(leftEntity, leftIndex);
+    }
+
+    public Object getRightValue() {
+        return variableDescriptor.getElement(rightEntity, rightIndex);
+    }
+
+    // ************************************************************************
+    // Worker methods
+    // ************************************************************************
+
+    @Override
+    public boolean isMoveDoable(ScoreDirector<Solution_> scoreDirector) {
+        // TODO maybe do not generate such moves
+        // Do not use Object#equals on user-provided domain objects. Relying on user's implementation of Object#equals
```

Comment:
(not directly related to this PR)
Existing moves (e.g. `SwapMove` and `ChangeMove`) do use `Object.equals()`. Something we may want to review?

Reply:
Yes. https://issues.redhat.com/browse/PLANNER-2613.